powershell-adapter/powershell.resource.ps1

```diff
@@ -1,427 +1,349 @@
-    }
-
-    return $false
+# format expected for configuration and resource output
+class configFormat {
+    [string] $name
+    [string] $type
+    [psobject] $properties
 }
 
-if (($PSVersionTable.PSVersion.Major -eq 7) -and ($PSVersionTable.PSVersion.Minor -eq 4) `
-   -and ($PSVersionTable.PSVersion.PreReleaseLabel.StartsWith("preview")))
-{
-    throw "PowerShell 7.4-previews are not supported by PowerShell adapter resource; please use PS 7.4.0-rc.1 or newer."
+# output format for resource list
+class resourceOutput {
+    [string] $type
+    [string] $kind
+    [string] $version
+    [string[]] $capabilities
+    [string] $path
+    [string] $directory
+    [string] $implementedAs
+    [string] $author
+    [string[]] $properties
+    [string] $requires
+    [string] $description
 }
 
-$inputobj_pscustomobj = $null
-if ($stdinput)
-{
-    $inputobj_pscustomobj = $stdinput | ConvertFrom-Json
-    $new_psmodulepath = $inputobj_pscustomobj.psmodulepath
-    if ($new_psmodulepath)
-    {
-        $env:PSModulePath = $ExecutionContext.InvokeCommand.ExpandString($new_psmodulepath)
-    }
+# dsc resource type (settable clone)
+class dscResource {
+    [moduleType] $ImplementationDetail
+    [string] $ResourceType
+    [string] $Name
+    [string] $FriendlyName
+    [string] $Module
+    [string] $ModuleName
+    [string] $Version
+    [string] $Path
+    [string] $ParentPath
+    [string] $ImplementedAs
+    [string] $CompanyName
+    [psobject[]] $Properties
 }
 
-$DscModule = Get-Module -Name PSDesiredStateConfiguration -ListAvailable |
-    Sort-Object -Property Version -Descending |
-    Select-Object -First 1
-
-if ($null -eq $DscModule)
-{
-    Write-Error "Could not find and import the PSDesiredStateConfiguration module."
-    # Missing module is okay for listing resources
-    if ($Operation -eq 'List') { exit 0 }
-
-    exit 1
+# module types
+enum moduleType {
+    ScriptBased
+    ClassBased
 }
 
-Import-Module $DscModule -DisableNameChecking
-
-# Adding some debug info to STDERR
-$m = gmo PSDesiredStateConfiguration
-$trace = @{"Debug"="PSVersion="+$PSVersionTable.PSVersion.ToString()} | ConvertTo-Json -Compress
-$host.ui.WriteErrorLine($trace)
-$trace = @{"Debug"="PSPath="+$PSHome} | ConvertTo-Json -Compress
-$host.ui.WriteErrorLine($trace)
-$trace = @{"Debug"="ModuleVersion="+$m.Version.ToString()} | ConvertTo-Json -Compress
-$host.ui.WriteErrorLine($trace)
-$trace = @{"Debug"="ModulePath="+$m.Path} | ConvertTo-Json -Compress
-$host.ui.WriteErrorLine($trace)
-$trace = @{"Debug"="PSModulePath="+$env:PSModulePath} | ConvertTo-Json -Compress
-$host.ui.WriteErrorLine($trace)
-
-if ($Operation -eq 'List')
-{
-    $DscResources= Get-DscResource
-
-    foreach ($r in $DscResources)
-    {
-        if ($r.ImplementedAs -eq "Binary")
-        {
+# Cache the results of Get-DscResource to optimize performance
+function Invoke-CacheRefresh {
+    # cache the results of Get-DscResource
+    [resourceCache[]]$resourceCache = @()
```

Comment:
Created https://github.com/PowerShell/DSC/issues/371 to persist cache to disk in the future